### ISSUE TITLE
Sensu-server and their handlers instrumentation

### ISSUE THREAD

**Issue by jaimegago** [sensu/sensu#1027]
At the moment there is no easy way (e.g. API) to list all sensu-servers of an install and their status, in particular in terms of handlers (as seen in the server logs "unknown handler: foo").
Checking on the handlers is a requirement since _all_ sensu servers must be configured identically because the balancing being done at the transport level (results and keepalive).

Sensu servers and handlers instrumentation (ideally an API) would be -extremely- useful to:
- spot rogue sensu-servers
- spot misconfigured sensu-servers

Both situations potentially leading to very erratic Sensu behavior (e.g. missing alerts, metrics being dropped to the floor, etc.).

FWIW we were testing our Sensu config mgmt automation via Vagrant boxes when one of them was connected to a live sensu cluster by accident and started the mess, that was a PITA to troubleshoot. It was picking from the results queue but not able to act (i.e. shipping metrics to graphite)  for connectivity reasons, tracking it down would have been greatly facilitated by some instrumentation of the sensu servers at the "sensu cluster" level.

**Comment by jaimegago** [comment]
This is very much related to https://github.com/sensu/uchiwa/issues/357

**Comment by portertech** [comment]
Hey @jaimegago, we can look into storing some dynamic Sensu server info in Redis, such as loaded config SHAs, and make it available via the API /info endpoint.

**Comment by jaimegago** [comment]
@portertech That'd be awesome and would definitively allow to track rogue and/or misconfigured sensu servers. If I understand this correctly it would enable dashboards to list all sensu servers of a sensu cluster and display misconfigured ones, right?

Now what would be even more awesome (and maybe it should be another issue/feature request) is to have the same information we have for failing checks (as in "can't run this command" failure) for the server handlers available so in addition to misconfigured at the **sensu** level (e.g. missing handler foo config that others in the cluster have) we can track misconfigured at the OS level that impact sensu, e.g. all sensu servers are configured correctly but server "foo" can't connect to "bar" and as such is not "sensu server healthy", or "all sensu servers are configured to run handler "foo" but server baz doesn't  have the actual handler script available at the specified path. Correct me if I'm wrong but this wouldn't be caught by the config SHAs, right?
At the moment unless -I'm missing something- the only way to get handlers failures is via the logs or is that in Redis already?

**Comment by jaimegago** [comment]
@portertech Any progress on this?

**Comment by calebhailey** [comment]
@jaimegago - interesting issue, and there are some good ideas in here, but a solution isn't as obvious... 

Storing a config SHA per Sensu server does sound interesting, though definitely doesn't account for whether the actual handler executables themselves are installed and/or are the same/correct version. Having said that, although there are a lot of seemingly obvious solutions, they all have their own pros/cons - and most of the cons would be _very_ misleading (as you suggested), potentially giving an operator the false-positive indication that "everything is green", when in fact there could be significant configuration drift between the systems. 

Some quick and dirty examples from my very [simple minded perspective][1]:

- "Just hash the config + handlers directory" wouldn't work because not all users install handler executables in the same location, plus many handler scripts have system dependencies, so verifying that the handler scripts are installed is only half of the battle. 
- "Just hash the config + all the configured handler `command` scripts" wouldn't work because you'd have to parse the `command` attribute for the actual command executable (which would be error prone, because `command` attributes accept command arguments, etc... too many variables), and then look for the executable on the local system. This could potentially work if an additional attribute (or set of attributes) were provided with the location(s) where Sensu should expect the corresponding command executable to be installed, but that adds a lot of config maintenance overhead and feels very rigid. 
- Any solution involving config hashing would be easy & fast, but hashing a potentially large number of handler scripts could be slow - and the most obvious place/time to do this (to me) would be when the Sensu server starts. That wouldn't be so bad if you never needed to restart Sensu, but restarts are required to pick up config changes, so this could make restarts really slow, which would be non-ideal.

Historically we've always punted on this issue and suggested that it should be solved via configuration management or some other automation solution (i.e. outside of Sensu), and _this is still true_, but I do think think there's a middle ground where there may be an opportunity to implement some simple functionality that help avoid a lot of the "gotchas" involved here. It's just a matter of figuring out an improvement that doesn't make things worse by providing a false-positive sense of security. :) 

In related news, we did just release a [`--print_config` CLI option][2] in Sensu 0.23, which prints Sensu's compiled configuration to STDOUT. It occurs to me that you may be able to leverage this configuration output to implement some of the ideas mentioned above as infrastructure tests in your provisioning pipeline. For example, you could try adding a `"verify_handler_deps": []` attribute to all handler definitions with paths to handler dependencies you want to verify/hash for each handler, and write a infrastructure test that hashes Sensu's compiled config + looks up and verifies/hashes all defined `verify_handler_deps` files, and compares the result with your expected result. 

[1]: https://img.buzzfeed.com/buzzfeed-static/static/2014-10/26/6/enhanced/webdr08/enhanced-14836-1414320930-8.jpg
[2]: https://sensuapp.org/docs/0.23/configuration#sensu-service-cli-arguments

**Comment by calebhailey** [comment]
@jaimegago I'm curious to get your thoughts on my previous comment. I wonder if we can close this issue as a "won't fix" for now for reasons mentioned above (i.e. this is a problem to solve outside of Sensu), unless you have some ideas about what we could do in Sensu that wouldn't mislead users.

**Comment by jaimegago** [comment]
@calebhailey  
Disclaimer: 3+ years into Sensu and I still have some gray areas, I am no rubbyist and have only contributed plugins, nothing to the core, so don't be surprised if there are errors in my reasoning here.

I do think there is a reasonable middle ground for Sensu servers instrumentation that I'd hope you guys can add.
I assume somewhere in Sensu (Redis?) there is a *live* record of the Sensu servers, just exposing that via an API would be great as it would allow to have programatic monitoring for rogue servers by cross comparing with config mgmt automation. It'd also be a simple "eye ball" verification tool (e.g. wait what's this "foobar" sensu server in my cluster?). This doesn't seem like a big feature to add and if anything please get us that =).

Now as far as providing Sensu server config validation how about looking at this from a Sensu server health state perspective? If a server picks a message that it can not handle (pun intended) it will yell at you in the logs to me this means the Sensu server has some health state awareness, that state should IMHO be tracking all handlers erroring and be reported via the health endpoint, then it is up to the admins to figure out where the problem lies (from connectivity issues to configuration ones).
In short there are logs when the handler isn't defined, I propose to extend it to all failing handlers executions.
Indeed if one of my Sensu servers handler execution is failing it is definitively not healthy and a potential source of havoc. Basically if a Sensu server picks an event and doesn't successfully handles it then it's unhealthy, if all events are handled successfully then it's green solid.
So my proposal is not so much Sensu Servers configuration validation as it is for a real Sensu server health status based on Sensu servers event handling instrumentation.

**Comment by calebhailey** [comment]
@jaimegago ah, now these are totally feasible ideas. There are a few related issues floating around too that I'll link as soon as I get a chance. For now I'm gonna tag this for consideration in 0.25. I'll update this again as we have some ideas about exactly what improvements we'll be shipping.

**Comment by calebhailey** [comment]
The 0.25 release is going to hit today with fewer features than originally anticipated & primarily only some internal improvements. Moving this issue to 0.26.

**Comment by portertech** [comment]
#1573 and #1628 implement most of the functionality described in this issue.

**Comment by portertech** [comment]
It's important to note that Sensu servers can still have "out-of-band" resources missing or misconfigured, for example, a missing pipe handler script/plugin or pipe handler runtime dependency or incorrect version. This Sensu server registry and settings hexdigest do however provide enough information to identify that their is a potential problem :+1: 

Considering this one closed :) (finally, thank you for your patience!)

**Comment by jaimegago** [comment]
Yey! Thx for following up, Sensu FTW!
